File: c++/src/line.cpp
#include "line.h"

Line::Line(Point& p1_, Point& p2_): p1(p1_), p2(p2_), angle(calculateAngle()){}


double Line::calculateAngle(){
    
    double dx = p2.x - p1.x;
    double dy = p2.y - p1.y;
    
    return atan2(dy,dx);
}

Point Line::getP1(){
    return p1;
}

Point Line::getP2(){
    return p2;
}

Angle Line::getAngle(){
    return angle;
}


Point* Line::midPoint(){
    double px = (p2.x+p1.x)/2;
    double py = (p2.y+p1.y)/2;

    return new Point(px,py);
}

bool Line::checkOnLine(Point& p){

    double dx = p.x - p1.x;
    double dy = p.y - p1.y;

    double temp = atan2(dy,dx);

    //DEBUG_MSG_L("ANGLE: " << temp << "\t" << angle);

    // check if the angle of p1 and p is equal
    if(angle == atan2(dy,dx)){

        // CHECK DISTANCES

        double distance = p.distance(p1);
        double lineLength = p2.distance(p1);

        return (distance <= lineLength);
        
    }

    return false;
}


/*
Check if line lies on each side of the input ray
*/
bool Line::checkEndPointIntersection(Point& p, Angle& a){
    
    // create two temporary points for the line endpoints
    Point temp1(p1.x,p1.y);
    Point temp2(p2.x,p2.y);

    temp1.translate(-p.x,-p.y);
    temp2.translate(-p.x,-p.y);

    DEBUG_MSG_L(temp1 << "\t" << temp2);

    // rotate the temp points to match the ray
    Angle reverse(-a.getAngle());
    temp1.rotate(reverse);
    temp2.rotate(reverse);
    
    DEBUG_MSG_L(temp1 << "\t" << temp2  << "\t" << reverse.degrees());
    
    // if the sign of each y component are the same, intersection is impossible
    return temp1.y * temp2.y <= 0.0f;
}


/*
Check if an input ray (point and angle) could intersect the line
*/
bool Line::checkPossibleIntersection(Point& p, Angle& a){

    // create two temporary points for the line endpoints
    Point temp1(p1.x,p1.y);
    Point temp2(p2.x,p2.y);

    temp1.translate(-p.x,-p.y);
    temp2.translate(-p.x,-p.y);
    DEBUG_MSG_L("TEMP1: " <<  temp1 << " TEMP2:" << temp2);

    // rotate the temp points to match the ray
    Angle reverse(-a.getAngle());
    temp1.rotate(reverse);
    temp2.rotate(reverse);

    DEBUG_MSG_L("TEMP1: " <<  temp1 << " TEMP2:" << temp2);
    // if the sign of each y component are the same, intersection is impossible
    if(temp1.y * temp2.y > 0.0f){
        return false;
    }

    /*
    X check
     - check the x normalized to y sum is greater than 0
     - the ray is currently at angle 0, so lines that have negative weight have an intersection point in -x
    */

    double normal1 = temp1.x/fabs(temp1.y);
    double normal2 = temp2.x/fabs(temp2.y);

    DEBUG_MSG_L("NORMAL1: " <<  normal1 << " NORMAL2:" << normal2);

    return (normal1 + normal2 > 0);

}


bool Line::checkPossibleIntersection(Line& line){

    Point temp = line.getP1();
    Angle angle = line.getAngle();

    // check if the ray condition holds for each line
    return checkPossibleIntersection(temp,angle) && line.checkPossibleIntersection(p1, angle);
}


// check if an input ray intersects with the line
Point* Line::intersection(Point& p, Angle& a){
        
    if(angle == a){
        return NULL;
    }
    
    double s1 = angle.tangent();
    double s2 = a.tangent();

    double V = (p1.y - p.y + s1 * (p.x - p1.x)) / (s2 - s1);

    double px = p.x + V;
    double py = p.y + s2 * V;

    Point temp(px,py);

    DEBUG_MSG_L(temp << checkOnLine(temp));

    return new Point(px,py);
    /*
    if(checkOnLine(temp)){
        return new Point(px,py);
    }
    else{
        return NULL;
    }
    */
}

// find the intersection point between this line and another line ~ return NULL if the intersection point is not on the lines
Point* Line::intersection(Line& line){
    
    // get the result if the input line is a ray

    Point p = line.getP1();
    Angle a = line.getAngle();

    Point* result = intersection(p,a);

    if(result){
        
        // if there is a point, and that point is not on the input line, then return NULL (no intersecting points)
        if(!line.checkOnLine(*result))
            result = NULL;
    }

    return result;
}

std::ostream& operator<<(std::ostream &strm, const Line &l){
    return strm << "LINE: " <<  l.p1 << " " << l.p2 << " " << l.angle;   
}

File: python/geometry/line.py
'''
Line
This class represents a line (two points) or a ray (point and angle). Has logic to find intersections.

@author ejbosia
'''

import numpy as np
import math
import logging

from geometry.point import Point

logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)

class Line:

    def __init__(self, p1, p2=None, angle=None):
        
        self.p1 = p1

        if p2 is None and angle is None:
            raise Exception("P2 or Angle required")
        
        self.p2 = p2

        if angle is None:
            self.angle = self.__angle()
        else:
            self.angle = angle


    # find the angle of the line from p1 to p2
    def __angle(self):
        dx = self.p2.x - self.p1.x
        dy = self.p2.y - self.p1.y
        return np.arctan2(dy,dx)


    # find the slope of the line (dy/dx)
    def slope(self):

        dx = math.cos(self.angle)
        dy = math.sin(self.angle)

        return dx,dy


    # find the point at the center of the line
    def bisect(self):
        
        if self.p2 is None:
            return None

        x = (self.p1.x + self.p2.x)/2
        y = (self.p1.y + self.p2.y)/2

        return (x,y)


    # find the length of the line (ray has an infinite length)
    def length(self):
        if self.p2 is None:
            return np.inf

        dx = self.p2.x - self.p1.x
        dy = self.p2.y - self.p1.y

        return math.sqrt(dx**2 + dy**2)


    # traverse the line a distance from p1
    def traverse(self, distance):

        if distance > self.length():
            raise Exception("TRAVERSE TOO LONG")

        ux, uy = self.slope()

        px = distance * ux + self.p1.x
        py = distance * uy + self.p1.y

        return (px, py)


    '''
    Check if there is a possible intersection
    '''
    def check_possible_intersection(self, p, a):
        # check for parallel lines
        if((self.angle%np.pi) == (a%np.pi)):
            return False

        # create two temporary points for the line endpoints
        l1 = Point(self.p1.x, self.p1.y)
        l2 = Point(self.p2.x, self.p2.y)

        # move the points to 0
        l1.translate_XY(-p.x,-p.y)
        l2.translate_XY(-p.x,-p.y)

        # rotate the temp points to match the ray
        l1.rotate(-a)
        l2.rotate(-a)

        # if the sign of each y component are the same, intersection is impossible
        if(round(l1.y,10) * round(l2.y,10) > 0):
            return False

        '''
        X check
        - check the x normalized to y sum is greater than 0
        - the ray is currently at angle 0, so lines that have negative weight have an intersection point in -x
        '''

        if l1.y == 0:
            return l1.x > 0
        elif l2.y == 0:
            return l2.x > 0
        else:
            normal = l1.x/abs(l1.y) + l2.x/abs(l2.y)

            return normal > 0


    def intersection(self, line):

        check1 = self.check_possible_intersection(line.p1, line.angle)
        check2 = line.check_possible_intersection(self.p1, self.angle)

        if not check1 or not check2:
            return None
        
        s1 = np.tan(self.angle)
        s2 = np.tan(line.angle)

        V = (self.p1.y - line.p1.y + s1 * (line.p1.x - self.p1.x)) / (s2 - s1)

        px = line.p1.x + V
        py = line.p1.y + s2 * V

        return Point(px,py)



    
    # return the cross product of this and another line
    def cross_product(self, line):

        ax = self.p2.x-self.p1.x
        ay = self.p2.y-self.p1.y

        bx = line.p2.x-line.p1.x
        by = line.p2.y-line.p1.y

        result = (ax * by) - (ay*bx)

        return result
    

    # return a zip of the line for plotting
    def plot(self):
        
        if not self.p2 is None:
            X = [self.p1.x, self.p2.x]
            Y = [self.p1.y, self.p2.y]
            return X,Y

        else:
            raise Exception("NOT IMPLEMENTED YET")


    def __eq__(self, line):
        return line.p1 == self.p1 and line.p2 == self.p2

    def __repr__(self):
        return str(self.p1) + " - " + str(self.p2) + " ANGLE: " + str(round(self.angle,2))
